src/Accelerators/NNPA/Transform/ZHigh/ZHighClipToDLFloat.cpp

```diff
@@ -0,0 +1,134 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------- ZHighClipToDLFloat.cpp - ZHigh High Level Optimizer -------===//
+//
+// Copyright 2023- The IBM Research Authors.
+//
+// =============================================================================
+//
+// This file implements a set of rewritten rules to clip CPU numerical values
+// before passing to ZHighStick, which avoids data range violation error due to
+// the dlfloat range.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Conversion/LLVMCommon/ConversionTarget.h"
+#include "mlir/IR/Matchers.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Pass/Pass.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+#include "src/Accelerators/NNPA/Dialect/ZHigh/ZHighOps.hpp"
+#include "src/Accelerators/NNPA/Dialect/ZHigh/ZHighOps/OpHelper.hpp"
+#include "src/Accelerators/NNPA/Pass/NNPAPasses.hpp"
+#include "src/Dialect/ONNX/DialectBuilder.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+#include "src/Support/TypeUtilities.hpp"
+
+using namespace mlir;
+using namespace onnx_mlir;
+using namespace onnx_mlir::zhigh;
+
+namespace onnx_mlir {
+namespace zhigh {
+
+namespace {
+
+/// Check if a value is from or transitively from a zTensor without value
+/// modification.
+bool valueFromZTensor(Value tensor) {
+  if (!tensor.dyn_cast<BlockArgument>()) {
+    Operation *op = tensor.getDefiningOp();
+    // From a zTensor.
+    if (isa<ZHighUnstickOp>(op))
+      return true;
+    // Concat/Transpose does not change the input precision. So we can consider
+    // that the input is already in the dlfloat range if it comes from zTensor.
+    if (isa<ONNXConcatOp, ONNXTransposeOp>(op)) {
```

Comment:
There are many more that falls in that category: expand/flatten/gather/reshape/squeeze/unsqueeze/pad.
Then there are also ops that are out of bound friendly: Max (of any kinds), like ReduceMax, Relu... If their input where high enough, their output is guaranteed to be high enough too.